### DISCUSSION TITLE
How to get Unicode codepoint instead of glyph index after hb_shape?

### DISCUSSION THREAD

**Discussion by HamedJP** [harfbuzz/harfbuzz#4129] [Q&A] [answered]
I'm so sorry to ask this quastion here. I'm noob at github, so I don't know where to ask
before calling the hb_shpe, glyph_info[0].codepoint is unicode codepoint. and it is glyph index after hb_shape. After I set the buffer content type to HB_BUFFER_CONTENT_TYPE_UNICODE I still get glyph index!
isn't it supposed to give Unicode codepoint?
this is my code:
```
#include <iostream>
#include <harfbuzz/hb.h>

int main()
{
    hb_buffer_t *buf;
    buf = hb_buffer_create();
    hb_buffer_add_utf8(buf, "A", -1, 0, -1);

    hb_buffer_set_direction(buf, HB_DIRECTION_LTR);
    hb_buffer_set_script(buf, HB_SCRIPT_LATIN);
    hb_buffer_set_language(buf, hb_language_from_string("en", -1));

    hb_blob_t *blob = hb_blob_create_from_file("ARIALUNI.TTF");
    hb_face_t *face = hb_face_create(blob, 0);
    hb_font_t *font = hb_font_create(face);

    hb_buffer_content_type_t t1 = hb_buffer_get_content_type(buf);

    unsigned int glyph_count;
    hb_glyph_info_t *glyph_info = hb_buffer_get_glyph_infos(buf, &glyph_count);
    hb_glyph_position_t *glyph_pos = hb_buffer_get_glyph_positions(buf, &glyph_count);

    printf("Codepoint befor Shaping: %d\n", t1);
    printf("%x\n", glyph_info[0].codepoint);

    hb_shape(font, buf, NULL, 0);

    hb_buffer_content_type_t t2 = hb_buffer_get_content_type(buf);

    glyph_info = hb_buffer_get_glyph_infos(buf, &glyph_count);
    glyph_pos = hb_buffer_get_glyph_positions(buf, &glyph_count);

    printf("Codepoint after Shaping: %d\n", t2);
    printf("%x\n", glyph_info[0].codepoint);

    hb_buffer_set_content_type(buf, HB_BUFFER_CONTENT_TYPE_UNICODE);
    hb_buffer_content_type_t t3 = hb_buffer_get_content_type(buf);

    glyph_info = hb_buffer_get_glyph_infos(buf, &glyph_count);
    glyph_pos = hb_buffer_get_glyph_positions(buf, &glyph_count);

    printf("Codepoint after setting content type: %d\n", t3);
    printf("%x\n", glyph_info[0].codepoint);

    hb_buffer_destroy(buf);
    hb_font_destroy(font);
    hb_face_destroy(face);
    hb_blob_destroy(blob);
}
```
and this is my out put:
```
Codepoint befor Shaping: 1
41
Codepoint after Shaping: 2
24
Codepoint after setting content type: 1
24
```
How can I get 0x41 instead of 0x24?

**Comment by behdad** [comment, marked as answer]
You can't. The Unicode string is lost after shaping. Perhaps tell us what you are trying to accomplish?

**Reply by HamedJP** [reply]
یه کتابخونه هست به اسم imgui برای ساخت رابط کاربری ساده در بازی‌ها. با opengl یا چیزای شبیه به اون باید رندر کرد. متاسفانه فارسی رو پشتیبانی نمی‌کنه. می‌خواستم از کتابخونه شما ا برای یه پرژه شخصی استفاده کنم.
برای آدرس دهی به گلیف‌ها از کدپوینت (code-point) استفاده میکنه، و نه گلیف ایندکس (glyph index). دنبال راهی می‌گشتم که از کتابخونه شما خروجی کدپوینت بگیرم، که ظاهرا نمیشه
ببخشد که انگلیسی فارسی قاطی شد. زبانم انقدر خوب نیست که منظورم رو با انگلیسی بیان کنم.
ابزار درجه یکی ساختید. پیش باد👏👏👏

**Reply by behdad** [reply]
راحت‌ترین راه اینه که از کتاب‌خانه‌ی زیر استفاده کنین:
https://github.com/HOST-Oman/libraqm

**Reply by HamedJP** [reply]
سپاس فراوان❤️❤️❤️
